GANDLF/utils/logger.py

```diff
@@ -0,0 +1,56 @@
+import logging, os, warnings
+from typing import Optional, Tuple
+from pathlib import Path
+
+
+def warning_on_one_line(message, category, filename, lineno, file=None, line=None):
+    """
+    This function formats warning message according to its type
+    """
+    if category == UserWarning:
+        return str(message)
+    else:
+        return "%s:%s: %s:%s" % (filename, lineno, category.__name__, message)
+
+
+def setup_logger(
+    output_dir: str, verbose: Optional[bool] = False
+) -> Tuple[logging.Logger, str, str]:
+    """
+    This function setups a logger with severity level controlled by verbose parameter from a config file.
+
+    Args:
+        logger_name (str): Name for a logger
+        logs_dir (str): Output directory for log files.
+        verbose (Optional[bool], optional): Used to setup the logging level. Defaults to False.
+
+    Returns:
+        logger (logging.Logger)
+        logger_dir (str): directory for the logs
+        logger_name (str): name of the logger
+    """
+    logs_dir = f"{output_dir}/logs"
+    logger_name = "gandlf"
```

Comment:
Hey! I aggree that using root logger would be the most strraightforward option (I wanted to do it myself), but I was reluctant to write something that is against "best practices" :) I'll update the code asap!

Reply:
I also have a question :) What do you think is more developer-friendly / readable: configuring logger in a function, or having it a json format?  (https://docs.python.org/3/howto/logging-cookbook.html#custom-handling-of-levels)

Reply:
Definitely function:) JSON / ini config are even more mind-blowing and error prone... That's may be useful for a bigger production services, but in case of Gandlf easier is better

Reply:
@sarthakpati you mean to store necessary values in the config instead of hardcoding them in python code, right? smth like this
```
from config import log_fortmat, log_level

def setup_logging():
    logging.basicConfig(format=log_format, ...)
    ...
```

Reply:
Btw I'm not sure about that....
Usually logging is initialized at tier-0, before any other logic starts to be executed. The goal is to use a proper logging for  all the code, including config-parser: in other case if there are any errors while config parsing, they would not be handled / printed properly as logging is not configured still

Reply:
We don't use ConfigManager everywhere, say [`gandlf_constructCSV`](https://github.com/mlcommons/GaNDLF/blob/master/gandlf_constructCSV): no config parsing at all
So if we want to use the same logging setup for all the scripts/commands, the explicit `from utils.logger import setup_logging` is better

Also, it would be quite easy to integrate with CLI tool: a separate function can be called at the app start, independent of if we use config further or not

And, personally, I don't expect `configManager` to contain any logic except config managing. But why do you prefer to integrate logging to configManager?

Reply:
Understood. 

That does indeed make more sense! 😄

Reply:
What do you think about getting rid of `verbose` parameter in config manager?
The current setup was dictated also by the fact that I wanted to include `verbose`, and since config is not parsed everywhere, I couldn't use it at the very beginning. Ideally I'd do it as follows:
- get rid of `verbose` parameter and have a DEBUG level in a root logger
- have a logs path independent from `output_dir`
- call logger with .getLogger(__name__) in modules
- std.err setup to ERROR level (to avoid annoying repetitive warnings)
- have only INFO messages in std.out OR use prints in std.out (then we could use INFOs in a different way - see https://guicommits.com/how-to-log-in-python-like-a-pro/); In the second approach (prints to std.out + INFOs as a information to have more control over the flow) we could gradually add info's to the project that might help us in debugging.
Any thoughts? :)

Reply:
> > have a logs path independent from output_dir

> Where do you wanna to place them? `./gandlf.log` or kinda this?

This can actually go into `parameters["output_dir"]/logs`. To make it robust, we can do something like this: `log_dir = os.path.join(parameters.get("output_dir", ${some_meaningful_default}), "logs")`.


> > all logger with .getLogger(name) in modules

> It means that it would be `root` logger that defines all handlers, formatters, filters, etc, right? We create a new logger in each module, but all of them would process messages in the same way, right? (as well as third-party libraries - say, logs from pytorch would be processed by root logger in the same way)

This will need *extremely* robust developer notes and documentation, and associated PR template update to ensure that the reviewers/contributors have this in mind.


> > get rid of verbose parameter and have a DEBUG level in a root logger

> AFAIU you mean that root logger would be connected to file handler without any filters, i.e. we'd write everything to the log file, right?

I think this makes sense.


> > std.err setup to ERROR level (to avoid annoying repetitive warnings)

> Didn't get. Do you wanna restrict `std.err` so only ERROR logs can be passed there? Or do you mean to catch all ERROR messages and pass them to std.err instead of std.out? In particular, `tqdm` prints its progress to stderr. And these annoying warnings - IMO they're stil worth to be left in a common debug log, aren't they?

Can we somehow structure `tqdm` to output to std.out instead? And I feel as long as warnings are present in a separate file, users can always check and take appropriate action. IMO, filtering _some_ information from the main output (which is currently usually noise) does make sense.


> > have only INFO messages in std.out OR use prints in std.out (then we could use INFOs in a different way - see https://guicommits.com/how-to-log-in-python-like-a-pro/)

> First, do you mean to have ONLY `info` in stdout or `info` and higher (`warning`, `error`)? Normal filter behavior is to keep lvl+ messages, i.e. if you filter by `info` it would mean you receive `info`, `warning`, `error`.  Second, I don't see any advantages of `print` comparing to `logging.info`. Why might we want to display message in console, but hide it from debug log? 

I will vote for any standardized mechanism for structuring messages. Also, I don't see any advantages of `print` statements compared to using `logging.info`; at best, we want to display on both console and the log.

Reply:
> > > have a logs path independent from output_dir
> > Where do you wanna to place them? `./gandlf.log` or kinda this?
> This can actually go into `parameters["output_dir"]/logs`. To make it robust, we can do something like this: `log_dir = os.path.join(parameters.get("output_dir", ${some_meaningful_default}), "logs")`.

Sorry, didn't get. `parameters` - do you mean model config file? or CLI command parameters? Also we can think about a common gandlf config file / config setup (like `~/.gandlf_config.yaml`, but that's definitely out of this PR's scope.

Reply:
> Can we somehow structure `tqdm` to output to std.out instead?

Yep, there is [`file` param](https://tqdm.github.io/docs/tqdm/#tqdm-objects)  that can be set to stdout

Reply:
> And I feel as long as warnings are present in a separate file, users can always check and take appropriate action. IMO, filtering _some_ information from the main output (which is currently usually noise) does make sense.

Didn't get, sorry. What do you mean under "main output"? and what (and where) do you want to filer from it?

Reply:
Discussed everything together outside of PR.

1. log is splitted into four parts: debug log file, stdout, stderr, tqdm progress file
- debug log file catches everything `DEBUG+`. In particular, it should catch errors and warnings etc also (thus, both stdout and stderr content is also catched by debug log file)
- stdout displays `INFO` messages. Not lower, not higher
- stderr displays `WARNING+` messages.
- tqdm is redirected to tqdm progress (btw, how @sylwiamm ? `tqdm(.., file=tqdm_file_path)`? a separate logger for tqdm? filter that catches tqdm messages?)

2. Excessive warnings from third party libs make a mess.
- `y_pred contains classes not in y_true`: solve here https://github.com/mlcommons/GaNDLF/issues/848
- `The ``converters`` are currently experimental.` - either patch ACSConv, or hide with a separate filter on our logging side

3. debug log file is located somewhere like  `TMP_FOLDER` / `gandlf` / `gandlf.log` with file rotation. It may be difficult to print the specific path in case of any exception, so I propose to print it once on the command start (if it possible at all?)


Add plz anything if I forgot / understand in wrong way, fabs